hypervisor/src/vm.rs

```diff
@@ -184,4 +205,15 @@ pub trait Vm: Send + Sync {
     fn state(&self) -> Result<VmState>;
     /// Set the VM state
     fn set_state(&self, state: &VmState) -> Result<()>;
+    /// set vmmops interface
+    fn set_vmmops(&mut self, vmmops: Arc<dyn VmmOps>) -> vm::Result<()>;
+    /// get vmmops interface
+    fn get_vmmops(&self) -> Result<Arc<dyn VmmOps>>;
+}
+
+pub trait VmmOps: Send + Sync {
+    fn write_guest_mem(&self, buf: &[u8], gpa:u64,) -> vm::Result<usize>;
+    fn read_guest_mem(&self, buf: &mut [u8], gpa:u64,) -> vm::Result<usize>;
+    fn mmio_write(&self, addr: u64, data: &[u8]) -> bool;
+    fn mmio_read(&self, addr: u64, data: &mut [u8]) -> bool;
```

Comment:
I guess you need `io_{read/write}` version for port I/O.

Reply:
yes. I will add that too